src/plugins/timeseries-plugin.ts

```diff
@@ -0,0 +1,258 @@
+            this.renderEngine.parent.render();
+        }
+    }
+
+    handleMouseUp() {
+        this.interactionsEngine.clearCursor();
+    }
+
+    setData(data: TimeseriesPoint[]) {
+        this.data = data;
+
+        let min = Number.MAX_VALUE;
+        let max = Number.MIN_VALUE;
+        let first = Number.MAX_VALUE;
+        let last = Number.MIN_VALUE;
+
+        this.data.forEach(([timestamp, value]) => {
+            if (value < min) {
+                min = value;
+            }
+            if (value > max) {
+                max = value;
+            }
+
+            if (timestamp < first) {
+                first = timestamp;
+            }
+            if (timestamp > last) {
+                last = timestamp;
+            }
+        });
+
+        this.summary = {
+            min,
+            max,
+            first,
+            last,
+        };
+    }
+
+    calcRect(start: number, duration: number, isEnd: boolean) {
+        const w = duration * this.renderEngine.zoom;
+
+        return {
+            x: this.renderEngine.timeToPosition(start),
+            w: isEnd ? (w <= 0.1 ? 0.1 : w >= 3 ? w - 1 : w - w / 3) : w,
+        };
+    }
+
+    handleHover(region: HitRegion<number> | null) {
+        this.hoveredRegion = region;
+    }
+
+    override renderTooltip() {
+        if (this.hoveredRegion) {
+            const round = (value: number) => Math.round(value * 100) / 100;
+            const data = this.hoveredRegion.data as TimeseriesPointWithIndex;
+
+            this.renderEngine.renderTooltipFromData(
+                [
+                    { text: `Value: ${round(data[TimeseriesPointWithIndexX.value])}` },
+                    {
+                        text: `Timestamp: ${round(data[TimeseriesPointWithIndexX.timestamp])}ms`,
+                    },
+                ],
+                this.interactionsEngine.getGlobalMouse()
+            );
+            return true;
+        }
+        return false;
+    }
+
+    override render() {
+        if (this.data?.length === 0 || !this.summary) {
+            return;
+        }
+
+        const timestampStart = this.renderEngine.positionX;
+        const timestampEnd = this.renderEngine.positionX + this.renderEngine.getRealView();
+
+        const positionStart = this.renderEngine.timeToPosition(timestampStart);
+        const positionEnd = this.renderEngine.timeToPosition(timestampEnd);
+
+        const heightPerValueUnit =
+            (this.height - this.styles.padding) / Math.max(this.summary.max - this.summary.min, 1);
+
+        const normalizeValue = (value: number) => {
+            return this.height - value * heightPerValueUnit;
+        };
+
+        const convertDataPointToPointWithIndex = (
+            idx: number,
+            time: number,
+            value: number
+        ): TimeseriesPointWithIndex => {
+            return [idx, time, value, this.renderEngine.timeToPosition(time), normalizeValue(value)];
```

Comment:
I didn't see any reason to create so many new arrays. It's better to just process the points in a loop and immediately get the position for each point.
```ts
this.data.forEach(([time, value]) => {
  const pos = this.renderEngine.timeToPosition(time);
  const relativeHeight = this.height - value * heightPerValueUnit;

  this.renderEngine.ctx.lineTo(pos, relativeHeight);
})
```

Reply:
I create 1 array. I perform 2 enumerations.

Reply:
What I mean is that you can avoid creating an array of datapoints. This array is huge, and if we have, for example, 100k points, we should create an array with 100k arrays with 7 elements. It would take something like 0.5-1ms per frame, and we only have 16ms per frame for all of our calculations and drawing. I think it's too much for a simple chart. Also, you create the same values every time - it's redundant.

Reply:
i will try